pkg/chain/ethlike/value.go

```diff
@@ -0,0 +1,53 @@
+package ethlike
+
+import (
+	"fmt"
+	"math/big"
+	"regexp"
+	"strings"
+)
+
+var Units = map[string]float64{
+	"wei":   1, // default unit
+	"gwei":  1e9,
+	"ether": 1e18,
+	"celo":  1e18,
+}
+
+type Value struct {
+	*big.Int
+}
+
+// UnmarshalText is a function used to parse an ETH-like value.
+func (v *Value) UnmarshalText(text []byte) error {
+	re := regexp.MustCompile(`^(\d+[\.]?[\d]*)[ ]?([\w]*)$`)
+	matched := re.FindSubmatch(text)
+
+	if len(matched) != 3 {
+		return fmt.Errorf("failed to parse value: [%s]", text)
+	}
+
+	number, ok := new(big.Float).SetString(string(matched[1]))
+	if !ok {
+		return fmt.Errorf(
+			"failed to set float value from string [%s]",
+			string(matched[1]),
+		)
+	}
+
+	unit := matched[2]
+	if len(unit) == 0 {
+		unit = []byte("wei")
+	}
+
+	if factor, ok := Units[strings.ToLower(string(unit))]; ok {
+		number.Mul(number, big.NewFloat(factor))
+		v.Int, _ = number.Int(nil)
+		return nil
+	}
+
+	return fmt.Errorf(
+		"invalid unit: %s; please use one of: wei, Gwei, ether, CELO",
+		unit,
+	)
+}
```

Comment:
My point of view was a little bit different: the `ethlike.Value` type represents a generic wei-based value on any ETH-like platform. Both ETH and CELO are wei based. Also, it doesn't mean ETH and CELO should be mutually convertible, they just have to be convertible on wei. However, I agree we could improve readability by exposing `ethereum.Wei` and `celo.Wei` which will embed the `ethlike.Value` and this way we'll avoid mixing everything in one place. WDYT?

Reply:
`ethlike.Value` represents either a price or balance. The price of 1 USD is something completely different than the price of 1 PLN. 1 USD account balance is something completely different than 1 PLN account balance. Similarly, CELO and Ether are two distinct currencies. Even though USD and PLN have the same precision (1 USD = 100 cents and 1 PLN = 100 groszy), 1 USD is something completely different than 1 PLN and 1 CELO is something completely different than 1 Ether.

What `ethlike.Value` can represent is 1e18 precision token with a function allowing to parse string in `<amount> <currency>` format into `(*big.Int, string)`, but IMO that's all. I'd recommend renaming `ethlike.Value` to `ethlike.Token` - this will be 1e18 precision token with some auxiliary parsing function inside, but `ethereum.Wei` and `celo.Wei` need to define their denominations separately.

Side note, Celo docs [say that `1 CELO = 10^18 CELO Wei`](https://docs.celo.org/celo-owner-guide/eth-recovery) so I do wonder if we shouldn't be having CELO Wei, CELO Gwei currency denomination codes for `celo.*` piece.

Reply:
> ethlike.Value represents either a price or balance. The price of 1 USD is something completely different than the price of 1 PLN. 1 USD account balance is something completely different than 1 PLN account balance. Similarly, CELO and Ether are two distinct currencies. Even though USD and PLN have the same precision (1 USD = 100 cents and 1 PLN = 100 groszy), 1 USD is something completely different than 1 PLN and 1 CELO is something completely different than 1 Ether.

Totally agree and I never argued with that :) I just tried to represent the same concept which you named the `ethlike.Token` but used a bad name and was a bit lazy by putting denominations in one place. Will refactor according to your proposition.

Reply:
Please see https://github.com/keep-network/keep-common/commit/dcb69c902ea3db8e1e981278ed9d94c9a6fc1b87

Reply:
Looks 👌 💯